### PR TITLE
Remove `unsafe` from ReadToString

### PR DESCRIPTION
**Pull request by goffrie**
We can do everything needed with only safe code.

### CODE REVIEW

**Comment by goffrie** [comment]
Thanks, good point. I've adjusted the implementation to try to put the buffer back on error. This doesn't work if the reader gets dropped, but fixing that doesn't really seem worth it.

File: tokio/Cargo.toml [hunk 1]
```diff
@@ -121,7 +121,7 @@ default-features = false
 optional = true
 
 [dev-dependencies]
-tokio-test = { version = "0.2.0" }
+tokio-test = { version = "0.2.0", path = "../tokio-test" }
```
Comment (goffrie) [inline tokio/Cargo.toml:124]: Not sure if this is kosher, but a new version of tokio-test with the `read_error` mock hasn't been released to crates.io yet.

Comment (carllerche) [reply tokio/Cargo.toml:124]: It *should* be ok to keep path dev-deps. What would be the reason to avoid them?

Comment (goffrie) [reply tokio/Cargo.toml:124]: I can't think of any reason, just wondering why it isn't already this way.

**Pull request merged**